feat(stl): add support for bond visualization to STL export

TheRealSami opened the pull request:
# Description

This pull request enhances the STL export functionality in `openbabel` by visualizing the bonds between atoms.
Previously, the STL output represented only atoms as spheres. With this change:

- Bonds can be visualized as cylinders connecting the bonding atoms (`-xb` option),
- Bond radius and spacing can be adjusted (`--bond-radius`/`--bond-spacing` options), and
- Bonds can be coloured according to [CPK coloring](https://en.wikipedia.org/wiki/CPK_coloring) (`-xc` option).

TLDR; This PR adds support for the [Ball-and-stick model](https://en.wikipedia.org/wiki/Ball-and-stick_model) in the STL export. 

# Implementation notes

- Implementation is limited to `stlformat.cpp`
- Unfortunately, testing was only performed locally since no tests for the STL format were found :(
- Added options: `--bond-radius`, `--bond-spacing`
- Modified options:
  - `-xc` now colours atoms and bonds (if the bond visualization is enabled)
  - `-xs <scaling_factor>` scales atom and bond size (if the bond visualization is enabled)

# Examples

| Molecule | No bonds [1] | With bonds [2] | With bonds + colour [3] |
|:--------:|:------------:|:--------------:|:-----------------------:|
| FH       | <img width="464" height="391" alt="fh" src="https://github.com/user-attachments/assets/9d5584f6-cd9a-4686-ac6c-a4d0c1653f61" /> | <img width="527" height="250" alt="fh_bonds" src="https://github.com/user-attachments/assets/741052d9-1a1b-4adc-8e5d-af8252a86955" /> | <img width="946" height="439" alt="fh_bonds+color" src="https://github.com/user-attachments/assets/72bef647-1a6c-4232-9666-a040655b675d" /> |
| C2H4     | <img width="464" height="455" alt="c2h4" src="https://github.com/user-attachments/assets/aaee5e3c-62d9-4211-a6d4-2bcddf33b1db" /> | <img width="736" height="533" alt="c2h4_bonds" src="https://github.com/user-attachments/assets/2f9bd7ab-bc03-4716-a810-63dc1d2b2f8f" /> | <img width="926" height="647" alt="c2h4_bonds+color" src="https://github.com/user-attachments/assets/268b7e16-361e-440f-8c3a-dd6189a5639c" /> |
| [Caffeine](https://pubchem.ncbi.nlm.nih.gov/compound/Caffeine) | <img width="621" height="541" alt="caffeine" src="https://github.com/user-attachments/assets/f045dedd-f06f-466f-b374-25cffc867ec6" /> |  <img width="682" height="612" alt="caffeine_bonds" src="https://github.com/user-attachments/assets/a16fc341-eb8b-404e-ab32-2f454c69388c" /> | <img width="607" height="535" alt="caffeine_bonds+color" src="https://github.com/user-attachments/assets/864d9beb-9c66-47bf-bef0-9a2283976bc8" /> |

> Examples are screenshots of  [viewstl.com](https://www.viewstl.com/#!)'s render of the models (unfortunately, wihtout any anti-aliasing)

[1]: No additional options
[2]: Options `-xs 0.2 -xb`
[3]: Options `-xs 0.2 -xb -xc`

Please let me know if there are any changes or additions you'd recommend!

TheRealSami committed:
fix(stl): set scale factor via the `-s` option

TheRealSami committed:
refactor(stl): modernize STLFormat source code

- Replace `std::vector<T>::push_back` with `std::vector<T>::emplace_back` to avoid temporary copies
- Replace preprocessor macro for colour with template function
- Define constants as `static constexpr`
- Replace C-style casts
- Declare local helper function as `static`
- Split up code for writing STL to `std::ostream`

TheRealSami committed:
feat(stl): add visualization for bonds

By providing the `-xb` option `obabel` visualizes the bonds between the atoms in the output STL.
The bond visualization can be parameterized by

- the scale factor (`-xs` option),
- the bond radius (`--bond-radius` option),
- the bond spacing - only applicable for bond orders > 1 (`--bond-spacing` option), and
- by enabling the colour export with the `-xc` option.

ghutchis commented:
Wow, looks great! As someone who has gotten into 3D printing molecular models, this would greatly simplify the workflow. As is, I've used VRML, imported into Blender then saved as STL in Blender for printing.

ghutchis commented:
BTW for colors, you should use the element table:
```
        double r, g, b;
        OBElements::GetRGB(atom->GetAtomicNum(), &r, &g, &b);
```

TheRealSami committed:
feat(stl)!: improve colour export for STL format

- Replace custom colour scheme for STL export with `openbabel`'s default colour scheme provided by `OBElements::GetRGB`
- Add `--color-order-rgb` flag to support both known - see [wikipedia stl](https://en.wikipedia.org/wiki/STL_(file_format)) - color channel orderings in the STL export

TheRealSami commented:
Hey @ghutchis thanks for the feedback, I'm happy that this useful to someone else as well! :)
As for the colours: commit https://github.com/openbabel/openbabel/pull/2814/commits/f451be18cfbfdd363563d93bc83fc48a7371929e now  uses `OBElements::GetRGB` rather than assigning custom colours for the STL export.

Also i've found out that my test website [viewstl.com](https://www.viewstl.com/#!) swaps the red and blue colour channels which leads to weird colours. Apparently this is an inherent problem with the STL file format since the color encoding isn't standardized - see [wikipedia stl file format](https://en.wikipedia.org/wiki/STL_(file_format)).
Since there is no way to automatically detect the correct colour channel ordering i also added the `--colour-order-rgb` flag to enable support for both RGB and BGR colour channel ordering with BGR being the default.

| Caffeine with colour ordering RGB | Caffeine with colour ordering BGR |
|:--------------------------------------------------:|:--------------------------------------------------:|
| <img width="745" height="677" alt="caffeine_color_ordering_rgb" src="https://github.com/user-attachments/assets/2347ec5c-d0e0-4af3-aaed-f9723a60f0b3" /> | <img width="745" height="677" alt="caffeine_color_ordering_bgr" src="https://github.com/user-attachments/assets/391af1a9-f7d7-49b1-8392-35827b027a3d" /> |

> Note: According to `openbabel`'s default colour scheme Oxygen should be red-ish and Nitrogen should be blue-ish

TheRealSami committed:
fix(stl): replace designated initializer list in bond modelling

Designated initializer lists are a C++20 features and therefore not compatible with the C++11 standard mandated by `openbabel`.
See [cppreference designated initializers](https://cppreference.com/w/cpp/language/aggregate_initialization.html#Designated_initializers) for more information.

ghutchis commented:
Thanks. We will likely switch to C++17 after the next release, but it'll be a bit before we move to C++20.

There seems to be an un-related test failure that I'll investigate.

fredrikw commented:
Looks great! Shouldn't the new options be Write options for consistency? And maybe shorter to match all the other format specific options.

TheRealSami commented:
Hey @fredrikw , thanks for the feedback!
Yes, you are right ideally the options would be defined as write options.
Unfortunately, `obabel` has some technical limitations making this tricky in practice:

1. (write) options are registered globally for all formats, which causes warnings at runtime if the same option is registered with differing argument counts, and
    - References: https://github.com/openbabel/openbabel/issues/170 , [obconversion.cpp#L1675](https://github.com/openbabel/openbabel/blob/master/src/obconversion.cpp#L1675)
2. write/ options are limited to single characters
    - Reference: [obabel.cpp#L266](https://github.com/openbabel/openbabel/blob/master/tools/obabel.cpp#L266)

Because of this, I went with longer, format-specific option names defined directly in `stlformat.cpp`, since it felt more user-friendly and maintainable.

That said, I'm open to adjusting the option names if you have suggestions that would better balance consistency with practicality. :)
Happy to update if there's a preferred convention here!

ghutchis reviewed (approved):
This looks good - sorry for the long delay - I haven't had much time I could devote to openbabel.
